### PR TITLE
update schema to reflect aliasing behavior

### LINKED ISSUE #38555: Inconsistent behavior of .reshape() and .flatten() in JIT
## 🐛 Bug

The behavior of `.reshape()` and `.flatten()` differ in pytorch and jit.

## To Reproduce

If we run the following code:

```
import torch


def reshape_test(t: torch.Tensor):
    t.reshape(-1)[0] = torch.tensor(1, dtype=t.dtype, device=t.device)

    return t


t = torch.zeros([2, 2])
print('PyTorch function result')
print(reshape_test(t))
print()

reshape_test = torch.jit.script(reshape_test)
t = torch.zeros([2, 2])
print('TorchScript function result')
print(reshape_test(t))
print()

print('TorchScript function graph')
print(reshape_test.graph)

```

We will get:
```
PyTorch function result
tensor([[1., 0.],
        [0., 0.]])

TorchScript function result
tensor([[0., 0.],
        [0., 0.]])

TorchScript function graph
graph(%t.1 : Tensor):
  return (%t.1)

```
The same behavior holds true if we change `.reshape(-1)` to `.flatten()`, whereas `.view(-1)` works as expected.

## Expected behavior

The results should be consistent or at least there should be some warnings that they can be inconsistent.

## Environment
PyTorch version: 1.6.0.dev20200515

cc @ezyang @gchanan @zou3519 @suo

### PR DESCRIPTION
**Pull request by robieta**
Fixes: https://github.com/pytorch/pytorch/issues/38555

I did an audit of `native_functions.yaml` and found several functions in addition to `reshape` which were not reporting that they could alias:

```
@torch.jit.script
def foo(t: torch.Tensor):
    new_value = torch.tensor(1, dtype=t.dtype, device=t.device)

    t.flatten()[0] = new_value
    t.reshape(-1)[1] = new_value
    t.view_as(t)[2] = new_value
    t.expand_as(t)[3] = new_value
    t.reshape_as(t)[4] = new_value
    t.contiguous()[5] = new_value
    t.detach()[6] = new_value

    return t
```

Currently none of the values are assigned after dead code elimination, after this PR all are. (And the JIT output matches that of eager.)

I don't think this needs to be unit tested; presumably the generic machinery already is and this just brings these ops under the same umbrella.

**BC-breaking note**: This updates the native operator schema and the aliasing rules for autograd. JIT passes will no longer incorrectly optimize mutations on graphs containing these ops, and inplace ops on the result of `flatten` will now properly be tracked in Autograd and the proper backward graph will be created.

### CODE REVIEW

**Comment by suo** [comment]
What are the changes in the serialized code? I'm sort of surprised there are any, perhaps it's due to different optimizations occurring.

Otherwise, this looks fine from a BC perspective, adding alias annotations shouldn't really be considered a BC break because it does not cause any existing usages of a schema to fail. You can whitelist it

File: tools/autograd/gen_autograd.py [hunk 1]
```diff
@@ -42,21 +42,28 @@
...
     'unsqueeze': 'self',
+    'flatten': 'flatten',
     'view': 'self',
+    'view_as': 'self',
     'unbind': 'self',
     '_indices': 'self',
     '_values': 'self',
     'indices': 'self',
     'values': 'self',
+    'contiguous': 'self',
```
Comment (albanD) [inline tools/autograd/gen_autograd.py:65]: Can you move this and the reshape to the list below please.
These are composite ops that should not be in this list.

File: tools/autograd/gen_autograd.py [hunk 2]
```diff
@@ -42,21 +42,28 @@
...
     'select': 'self',
     'slice': 'self',
     'squeeze': 'self',
     't': 'self',
     'transpose': 'self',
     'unfold': 'self',
     'unsqueeze': 'self',
+    'flatten': 'flatten',
     'view': 'self',
+    'view_as': 'self',
```
Comment (albanD) [inline tools/autograd/gen_autograd.py:59]: This one is also a composite op.

File: tools/autograd/gen_autograd.py [hunk 3]
```diff
@@ -42,21 +42,28 @@
     'as_strided': 'self',
     'diagonal': 'self',
     'expand': 'self',
+    'expand_as': 'self',
```
Comment (albanD) [inline tools/autograd/gen_autograd.py:45]: This is a composite op as well.

File: tools/autograd/gen_autograd.py [hunk 4]
```diff
@@ -42,21 +42,28 @@
...
+    'flatten': 'flatten',
     'view': 'self',
+    'view_as': 'self',
     'unbind': 'self',
     '_indices': 'self',
     '_values': 'self',
     'indices': 'self',
     'values': 'self',
+    'contiguous': 'self',
+    'pin_memory': 'pin_memory',
```
Comment (albanD) [inline tools/autograd/gen_autograd.py:66]: This one does not always returns a view. So it should not be here I'm afraid.
I think we want to make it composite and add here only the function corresponding to the case where it is actually a view.

File: tools/autograd/gen_autograd.py [hunk 5]
```diff
@@ -42,21 +42,28 @@
...
+    'reshape': 'self',
+    'reshape_as': 'self',
     'select': 'self',
     'slice': 'self',
     'squeeze': 'self',
     't': 'self',
     'transpose': 'self',
     'unfold': 'self',
     'unsqueeze': 'self',
+    'flatten': 'flatten',
```
Comment (albanD) [inline tools/autograd/gen_autograd.py:57]: This is a good catch. Be sure to add the PR as BC-breaking and add a note in the PR description stating that `flatten` will now properly detect it's output as a view. So inplace ops on the result will now properly be tracked and the proper backward graph will be created.

**Comment by robieta** [comment]
> What are the changes in the serialized code? I'm sort of surprised there are any, perhaps it's due to different optimizations occurring.
> 
> Otherwise, this looks fine from a BC perspective, adding alias annotations shouldn't really be considered a BC break because it does not cause any existing usages of a schema to fail. You can whitelist it

Ah, it's just bumping a version counter since this technically changes the operator schema. The rest of the files are identical.

File: tools/autograd/gen_autograd.py (hunk 1 above)
Comment (robieta) [reply tools/autograd/gen_autograd.py]: Done.

File: tools/autograd/gen_autograd.py (hunk 2 above)
Comment (robieta) [reply tools/autograd/gen_autograd.py]: Done.

File: tools/autograd/gen_autograd.py (hunk 3 above)
Comment (robieta) [reply tools/autograd/gen_autograd.py]: Done.

File: tools/autograd/gen_autograd.py (hunk 4 above)
Comment (robieta) [reply tools/autograd/gen_autograd.py]: It's not clear if that means `pin_memory` should be in the second list or omitted entirely.

File: tools/autograd/gen_autograd.py (hunk 5 above)
Comment (robieta) [reply tools/autograd/gen_autograd.py:55]: Done, thanks.

**Comment by robieta** [comment]
Thanks for the review! This has been very informative.

File: tools/autograd/gen_autograd.py [hunk 6]
```diff
@@ -44,13 +44,15 @@
     'expand': 'self',
     'narrow': 'self',
     'permute': 'self',
+    'reshape_as': 'self',
```
Comment (albanD) [inline tools/autograd/gen_autograd.py:47]: I think reshape_as is the same as reshape here and should be moved below.

File: tools/autograd/gen_autograd.py [hunk 7]
```diff
@@ -44,13 +44,15 @@
...
     'select': 'self',
     'slice': 'self',
     'squeeze': 'self',
     't': 'self',
     'transpose': 'self',
     'unfold': 'self',
     'unsqueeze': 'self',
+    'flatten': 'flatten',
```
Comment (albanD) [inline tools/autograd/gen_autograd.py:55]: The value for the key should be the argument that the output is a view of. Here it should be 'self' I think.

File: tools/autograd/gen_autograd.py [hunk 8]
```diff
@@ -68,7 +70,8 @@
 # this list contains both the root view functions and any that are purely composed
 # of viewing functions, and is used by the JIT to determine when an operator
 # returns a view of its inputs
```
Comment (albanD) [inline tools/autograd/gen_autograd.py:72]: Could you add a note in this comment that this link contains all the functions that **can** return a view they might not return views in some cases (like contiguous).

**Review by albanD — COMMENTED** [review]
Good for me except the two small details below.

File: tools/autograd/gen_autograd.py (hunk 6 above)
Comment (robieta) [reply tools/autograd/gen_autograd.py]: Done.

File: tools/autograd/gen_autograd.py (hunk 7 above)
Comment (robieta) [reply tools/autograd/gen_autograd.py]: Ooof. Good catch.

File: tools/autograd/gen_autograd.py (hunk 8 above)
Comment (robieta) [reply tools/autograd/gen_autograd.py]: Good point.

**Comment by albanD** [comment]
The autograd part is good for me. But you will need someone from the jit team to check the jit part.

**Comment by robieta** [comment]
> Thanks for making all the changes! The autograd part is good for me. But you will need someone from the jit team to check the jit part.

SGTM. I will block on Michael's review.

**Review by suo — APPROVED** [review]
Thanks for these changes!

**Comment by albanD** [comment]
Don't forget that you need to add these to the backward compatibility whitelist before landing  ;)

**Pull request closed without merging**